themes/default/content/blog/exploring-circular-dependencies/index.md

```diff
@@ -0,0 +1,226 @@
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: ping.name,
+    },
+  },
+});
+```
+
+Programming languages don’t let us create one object which takes a value from another object that doesn’t yet exist. Therefore, we can’t express a circular dependency in this way.
+
+The simplest workaround for this error is to get a reference to our own stack, which contains the outputs from the previous deployment. Now, we set the reference to `pong` within `ping` from the output of the last deployment.
+
+```typescript
+const currentStack = new pulumi.StackReference(
+  `acme-org/${pulumi.getProject()}/${pulumi.getStack()}`
+);
+
+const ping = new CallbackFunction("ping", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: currentStack
+        .getOutput("pongName")
+        .apply((pongName) => pongName ?? ""),
+    },
+  },
+});
+
+const pong = new CallbackFunction("pong", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: ping.name,
+    },
+  },
+});
+
+export const pingName = ping.name;
+export const pongName = pong.name;
+```
+
+The downside of this approach is that we have to deploy twice to reach the desired state because we need the results of the previous deployment, though this approach might be fine for simple one-off circumstances.
+
+How then could we make the process easier and clearer for users who need to resolve this kind of circular dependency regularly? We explored four options as part of the hackathon, so let’s go through each one, getting progressively closer to a possible solution.
+
+## Option 1: State
+
+Instead of referencing outputs directly from our stack’s previous deploy, and manually deploying twice, we could introduce a special type called “state”.
+
+The `state` object does 3 tasks:
+
+1. Read the value from the previous deployment, or provide a default if it’s the first deployment.
+2. Update the value in the stack state using a “set” method.
+3. Automatically triggers another deployment if its value has changed.
+
+```typescript
+const pongName = new pulumi.State("pong-name", {
+  initialValue: "",
+});
+
+const ping = new CallbackFunction("ping", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: pongName.value,
+    },
+  },
+});
+
+const pong = new CallbackFunction("pong", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: ping.name,
+    },
+  },
+});
+
+// Engine re-runs "up" if this changes the value
+pongName.set(pong.name);
+```
+
+Deployments should always match the preview, but when triggering the re-deployment, we’re making a change that wasn’t in the preview. This action could also result in a never-ending loop where every deployment makes another change to the state and triggers another deployment. So let’s explore a different option that might not have this issue.
```

Comment:
Maybe be more explicit here that only the first deployment will match the preview?